File: src/main/java/io/github/justfoxx/teacup/v1/registry/SetRegistry.java
package io.github.justfoxx.teacup.v1.registry;

import com.google.common.collect.ImmutableSet;
import org.jetbrains.annotations.NotNull;

import java.util.Iterator;
import java.util.Objects;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

public class SetRegistry<V> implements Iterable<V> {
    private final Set<V> registry = ConcurrentHashMap.newKeySet();
    
    public V add(V value) {
        Objects.requireNonNull(value);
        registry.add(value);
        return value;
    }
    
    public ImmutableSet<V> getAll() {
        return ImmutableSet.copyOf(registry);
    }

    @NotNull
    @Override
    public Iterator<V> iterator() {
        return registry.iterator();
    }
}

File: src/main/kotlin/io/github/justfoxx/teacup/v1/registry/SetRegistry.kt
package io.github.justfoxx.teacup.v1.registry

import com.google.common.collect.ImmutableSet

/**
 * A registry that stores values in only by value
</V> */
class SetRegistry<V> {
    private val registry = MapRegistry<V, Any>()

    /**
     * Adds a value to the registry
     * @param value The value to add
     * @return The value added
     */
    fun add(value: V): V {
        registry.add(value, Object())
        return value
    }

    val all: ImmutableSet<V>
        /**
         * Gets all the values in the registry
         * @return All the values in the registry
         */
        get() = ImmutableSet.copyOf(registry.all.map { it.key })
}